source/math/FloatConvexHull2D.ooc

```diff
@@ -33,42 +35,99 @@ FloatConvexHull2D: class {
 		newPoints := VectorList<FloatPoint2D> new(this count)
 		for (i in 0 .. this count)
 			newPoints add(transform * this _points[i])
 		This new(newPoints)
 	}
-	_isOnLeft: static func (leftPoint, rightPoint, queryPoint: FloatPoint2D) -> Bool {
-		(rightPoint x - leftPoint x) * (queryPoint y - leftPoint y) > (rightPoint y - leftPoint y) * (queryPoint x - leftPoint x)
+	computeHull: func {
+		if (this count > 2) {
+			// Uses the Quickhull algorithm if more than two points, average complexity O(n log n)
+			// http://www.cse.yorku.ca/~aaw/Hang/quick_points/Algorithm.html
+			points := this _points copy()
+			this _points clear()
+			leftMostIndex := 0
+			rightMostIndex := 0
+			for (i in 0 .. points count)
+				if (points[i] x < points[leftMostIndex] x)
+					leftMostIndex = i
+				else if (points[i] x > points[rightMostIndex] x)
+					rightMostIndex = i
+			leftEndpoint := points[leftMostIndex]
+			rightEndpoint := points[rightMostIndex]
+			leftSet := VectorList<FloatPoint2D> new()
+			rightSet := VectorList<FloatPoint2D> new()
+			for (i in 0 .. points count)
+				if (i != leftMostIndex && i != rightMostIndex)
+					if (This _isOnLeft(points[leftMostIndex], points[rightMostIndex], points[i]))
+						leftSet add(points[i])
+					else
+						rightSet add(points[i])
+			this _points add(leftEndpoint)
+			this _findHull(leftSet, leftEndpoint, rightEndpoint)
+			this _points add(rightEndpoint)
+			this _findHull(rightSet, rightEndpoint, leftEndpoint)
+			points free()
+			leftSet free()
+			rightSet free()
+		}
 	}
-	free: override func {
-		this _points free()
-		super()
+	_findHull: func (currentSet: VectorList<FloatPoint2D>, leftPoint, rightPoint: FloatPoint2D) {
+		if (currentSet count == 1)
+			this _points add(currentSet[0])
+		else if (currentSet count > 1) {
+			maximumDistance := Float negativeInfinity
+			maximumDistanceIndex := -1
+			for (i in 0 .. currentSet count) {
+				distance := This _pointsLinePseudoDistance(leftPoint, rightPoint, currentSet[i])
+				if (distance > maximumDistance) {
+					maximumDistance = distance
+					maximumDistanceIndex = i
+				}
+			}
+			maximumDistancePoint := currentSet[maximumDistanceIndex]
+			outsideLeft := VectorList<FloatPoint2D> new()
+			outsideRight := VectorList<FloatPoint2D> new()
+			for (i in 0 .. currentSet count)
+				if (i != maximumDistanceIndex)
+					if (This _isOnLeft(leftPoint, maximumDistancePoint, currentSet[i]))
+						outsideLeft add(currentSet[i])
+					else if (This _isOnLeft(maximumDistancePoint, rightPoint, currentSet[i]))
+						outsideRight add(currentSet[i])
+			this _findHull(outsideLeft, leftPoint, maximumDistancePoint)
+			this _points add(maximumDistancePoint)
+			this _findHull(outsideRight, maximumDistancePoint, rightPoint)
+			outsideLeft free()
+			outsideRight free()
+		}
+	}
+	_pointsLinePseudoDistance: static func (leftPoint, rightPoint, queryPoint: FloatPoint2D) -> Float {
```

Comment:
Also, if what it does is compute the distance between a point a line between two other points, it *may* belong in the `FloatPoint2D` class, should it be useful elsewhere.